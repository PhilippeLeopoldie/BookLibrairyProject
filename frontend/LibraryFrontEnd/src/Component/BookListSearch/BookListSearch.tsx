import { Book } from "../Book/Book";
import { useEffect, useState } from "react";
import url from "../../Url";

type BooksType = {
  book: {
    id: number;
    title: string;
    author: string;
  };
};

type BooksSearchCriteria = {
  searchCriteria?: {
    title: string;
    author: string;
  };
};

export const BookListSearch = ({ searchCriteria }: BooksSearchCriteria) => {
  const [books, setBooks] = useState<BooksType[]>([]);
  const [initialBooks, setInitialBooks] = useState<BooksType[]>([]);
  let filteredBooks: BooksType[] = [];
  const handleDeleteBook = (bookId: number) => {
    setBooks((books) => {
      if (books !== null) {
        return books.filter((book) => book.book.id !== bookId);
      } else {
        return [];
      }
    });
  };

  const fetchBooks = async () => {
    try {
      const booksResponse: Response = await fetch(url + "api/Book");
      if (booksResponse.status === 200) {
        const booksResponseData = await booksResponse.json();
        setInitialBooks(booksResponseData.$values);
      } else if (booksResponse.status === 404) {
        console.log(booksResponse);
      }
    } catch (error) {
      console.error("Error fetching books:", error);
    }
  };

  useEffect(() => {
    fetchBooks();
  }, []);

  useEffect(() => {
    if (searchCriteria !== undefined) {
      (searchCriteria.title === "" && searchCriteria.author === "") ? 
        filteredBooks = []
       : 
        filteredBooks = initialBooks.filter((books) => {
          const titleMatches = books.book.title
            .toLowerCase()
            .includes(searchCriteria.title.toLowerCase());

          const authorMatches = books.book.author
            .toLowerCase()
            .includes(searchCriteria.author.toLowerCase());

          return titleMatches || authorMatches;
        });
      
      setBooks(filteredBooks);
    }
  }, [searchCriteria]);

  if (!books) {
    return <h1>Loading...</h1>;
  }

  return (
    <>
      <div className="books">
        <h2>({books.length}) book(s) found</h2>
        <div className="booksContainer" data-testid="booksContainer">
          {books &&
            Array.isArray(books) &&
            books
              .sort((a, b) => b.book.id - a.book.id)
              .map((bookDetail, index) => (
                <Book
                  key={bookDetail.book.id}
                  book={bookDetail.book}
                  onDelete={handleDeleteBook}
                />
              ))}
        </div>
      </div>
    </>
  );
};
