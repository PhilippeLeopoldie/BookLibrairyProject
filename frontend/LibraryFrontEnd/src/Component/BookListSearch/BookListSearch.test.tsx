import {render, screen,} from "@testing-library/react";
import { BookListSearch } from "./BookListSearch";
import "@testing-library/jest-dom";
import { useState } from "react";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

type BooksType = {
  book: {
    id: number;
    title: string;
    author: string;
  };
};

const book : BooksType = {
  book:{
    id: 1,
    title: "title",
    author: "author",
  }

}

describe("BookListSearch", () => {
  test("snapShot: render BookListSearch", ()=> {
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);

    const {asFragment} = render(<BookListSearch titleOrAuthor={""}/>)

    expect(asFragment()).toMatchSnapshot();
  })

  test("snapShot: render 'Loading...'", ()=> {
    (useState as jest.Mock).mockImplementation(() => [null, jest.fn()]);
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);

    const {asFragment} = render(<BookListSearch titleOrAuthor={""}/>)

    expect(asFragment()).toMatchSnapshot();
  })
})