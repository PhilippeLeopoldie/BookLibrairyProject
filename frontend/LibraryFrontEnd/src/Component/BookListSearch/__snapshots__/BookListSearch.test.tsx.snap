// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListSearch snapShot: render 'Loading...' 1`] = `
<DocumentFragment>
  <div
    class="books"
  >
    <h2>
      () book(s) found
    </h2>
    <div
      class="booksContainer"
      data-testid="booksContainer"
    />
  </div>
</DocumentFragment>
`;

exports[`BookListSearch snapShot: render BookListSearch 1`] = `
<DocumentFragment>
  <div
    class="books"
  >
    <h2>
      () book(s) found
    </h2>
    <div
      class="booksContainer"
      data-testid="booksContainer"
    />
  </div>
</DocumentFragment>
`;
